libcxx/include/support/atomic/atomic_cuda.h

```diff
@@ -0,0 +1,488 @@
+#define ATOMIC_WCHAR_T_LOCK_FREE   2
+#define ATOMIC_SHORT_LOCK_FREE     2
+#define ATOMIC_INT_LOCK_FREE       2
+#define ATOMIC_LONG_LOCK_FREE      2
+#define ATOMIC_LLONG_LOCK_FREE     2
+#define ATOMIC_POINTER_LOCK_FREE   2
+#endif //!defined(__CLANG_ATOMIC_BOOL_LOCK_FREE) && !defined(__GCC_ATOMIC_BOOL_LOCK_FREE)
+
+#ifndef __ATOMIC_RELAXED
+#define __ATOMIC_RELAXED 0
+#define __ATOMIC_CONSUME 1
+#define __ATOMIC_ACQUIRE 2
+#define __ATOMIC_RELEASE 3
+#define __ATOMIC_ACQ_REL 4
+#define __ATOMIC_SEQ_CST 5
+#endif //__ATOMIC_RELAXED
+
+#ifndef __ATOMIC_BLOCK
+#define __ATOMIC_SYSTEM 0 // 0 indicates default
+#define __ATOMIC_DEVICE 1
+#define __ATOMIC_BLOCK 2
+#define __ATOMIC_THREAD 10
+#endif //__ATOMIC_BLOCK
+
+inline __host__ __device__ int __stronger_order_cuda(int __a, int __b) {
+    int const __max = __a > __b ? __a : __b;
+    if(__max != __ATOMIC_RELEASE)
+        return __max;
+    static int const __xform[] = {
+        __ATOMIC_RELEASE,
+        __ATOMIC_ACQ_REL,
+        __ATOMIC_ACQ_REL,
+        __ATOMIC_RELEASE };
+    return __xform[__a < __b ? __a : __b];
+}
+
+enum thread_scope {
+    thread_scope_system = __ATOMIC_SYSTEM,
+    thread_scope_device = __ATOMIC_DEVICE,
+    thread_scope_block = __ATOMIC_BLOCK,
+    thread_scope_thread = __ATOMIC_THREAD
+};
+
+#define _LIBCUDACXX_ATOMIC_SCOPE_TYPE ::cuda::thread_scope
+#define _LIBCUDACXX_ATOMIC_SCOPE_DEFAULT ::cuda::thread_scope::system
+
+struct __thread_scope_thread_tag { };
+struct __thread_scope_block_tag { };
+struct __thread_scope_device_tag { };
+struct __thread_scope_system_tag { };
+
+template<int _Scope>  struct __scope_enum_to_tag { };
+/* This would be the implementation once an actual thread-scope backend exists.
+template<> struct __scope_enum_to_tag<(int)thread_scope_thread> {
+    using type = __thread_scope_thread_tag; };
+Until then: */
+template<> struct __scope_enum_to_tag<(int)thread_scope_thread> {
+    using type = __thread_scope_block_tag; };
+template<> struct __scope_enum_to_tag<(int)thread_scope_block> {
+    using type = __thread_scope_block_tag; };
+template<> struct __scope_enum_to_tag<(int)thread_scope_device> {
+    using type = __thread_scope_device_tag; };
+template<> struct __scope_enum_to_tag<(int)thread_scope_system> {
+    using type = __thread_scope_system_tag; };
+
+template <int _Scope>
+_LIBCUDACXX_INLINE_VISIBILITY auto constexpr __scope_tag() ->
+        typename __scope_enum_to_tag<_Scope>::type {
+    return typename __scope_enum_to_tag<_Scope>::type();
+}
+// END TODO
+
+// Wrap host atomic implementations into a sub-namespace
+namespace host {
+#if defined(_LIBCUDACXX_COMPILER_MSVC)
+#  include "atomic_msvc.h"
+#elif defined (_LIBCUDACXX_HAS_GCC_ATOMIC_IMP)
+#  include "atomic_gcc.h"
+#elif defined (_LIBCUDACXX_HAS_C11_ATOMIC_IMP)
+//TODO
+// #  include "atomic_c11.h"
+#endif
+}
+
+#include "atomic_cuda_generated.h"
+#include "atomic_cuda_derived.h"
+
+template <typename _Tp>
+struct __skip_amt { enum {value = 1}; };
```

Comment:
"amt"? "amount"? I understand what this is for because I know the __atomic intrinsics, but I'd like this to have a better name. "__difference_scale" or something? (I don't think that's a good name either, but it's more evocative of what it is used for.)

Reply:
This is essentially the same hack that was in another layer. I'll try to use that one instead and get rid of the duplication.

Reply:
Fixed with a rename to __atomic_ptr_inc. I think that conveys the idea that it's an increment for pointers.